File: implementations/dotnet/Algorithms.Library/BipartiteGraph.cs
using System.Collections.Generic;
using System;

namespace Algorithms.Library
{
    public class BipartiteGraph
    {
        public bool IsBipartite(int[][] graph)
        {

            int[] colors = new int[graph.Length];
            int [,]matrix = new int[graph.Length,graph.Length];

            for (int i = 0; i < graph.Length; i++)
            {
                for (int j = 0; j < graph[i].Length; j++)
                {
                    matrix[i,graph[i][j]] =1;
                    matrix[graph[i][j],i] =1;

                }
            }
            Array.Fill(colors, -1);
            for (int i = 0; i < graph.Length; i++)
            {

                if (colors[i] == -1)
                {
                    colors[i] = 1;
                    if (!BFS(i, colors, matrix))
                        return false;
                }

            }

            return true;
        }

        private bool BFS(int source, int[] colors, int[,] graph)
        {

            Queue<int> queue = new();
            queue.Enqueue(source); //enqueueing the source

            while (queue.Count != 0)
            {
                int uSource = queue.Dequeue();



                for (int vecino = 0; vecino < graph.GetLength(0); vecino++)
                {
                    if (graph[uSource,vecino] == 1)
                    {
                        if (colors[vecino] == -1)
                        {
                            //coloreo con el color opuesto al mio;
                            colors[vecino] = 1 - colors[uSource];
                            queue.Enqueue(vecino);
                        }
                        else if (colors[vecino] == colors[uSource])
                            return false; // hay arista, pero el vecino esta coloreado como yo
                        else
                            continue;
                    }


                }
            }

            return true;

        }
    }
}

File: implementations/cpp/Algorithms.Executer/bipartitegraph.cpp
#include "bipartitegraph.h"

bool bipartitegraph::isBipartite(vector<vector<int>> &graph)
{

    vector<int> colors(graph.size(), -1);
    
    for (int i = 0; i < graph.size(); i++)
    {

        if (colors[i] == -1)
        {
            colors[i] = 1;
            if (!BFS(i, colors, graph))
                return false;
        }
    }

    return true;
}

bool bipartitegraph::BFS(int source, vector<int> colors, vector<vector<int>> &graph)
{

    queue<int> my_queue;
    my_queue.push(source); // enqueueing the source

    while (!my_queue.empty())
    {
        int uSource = my_queue.front();
        my_queue.pop();

        for (int vecino : graph[uSource])
        {
            

            if (colors[vecino] == -1)
            {
                // coloreo con el color opuesto al mio;
                colors[vecino] = 1 - colors[uSource];
                my_queue.push(vecino);
            }
            else if (colors[vecino] == colors[uSource])
                return false; // hay arista, pero el vecino esta coloreado como yo
            else
                continue;
        }
    }

    return true;
}